internal/btf/co_re.go

```diff
@@ -0,0 +1,761 @@
+		}
+
+		tacc := &coreAccessor{
+			typ: targetType,
+			idx: lacc.idx,
+		}
+		targetSpec.spec = append(targetSpec.spec, tacc)
+		targetSpec.rawSpec = append(targetSpec.rawSpec, tacc.idx)
+
+		sz, err := Sizeof(targetType)
+		if err != nil {
+			return nil, err
+		}
+		targetSpec.bitOffset += lacc.idx * uint32(sz) * 8
+	}
+
+	return targetSpec, nil
+}
+
+func areTypesCompatible(localType Type, targetType Type) (bool, error) {
+	if reflect.TypeOf(localType) != reflect.TypeOf(targetType) {
+		return false, nil
+	}
+
+	var err error
+	for depth := 0; depth <= maxTypeDepth; depth++ {
+		localType, err = skipModsAndTypedefs(localType)
+		if err != nil {
+			return false, err
+		}
+		targetType, err = skipModsAndTypedefs(targetType)
+		if err != nil {
+			return false, err
+		}
+
+		if reflect.TypeOf(localType) != reflect.TypeOf(targetType) {
+			return false, nil
+		}
+
+		switch v := localType.(type) {
+		case *Struct, *Union, *Enum, *Fwd:
+			return true, nil
+		case *Int:
+			tv := targetType.(*Int)
+			return v.Offset == 0 && tv.Offset == 0, nil
+		case *Pointer:
+			tv := targetType.(*Pointer)
+			localType = v.Target
+			targetType = tv.Target
+			continue
+		case *Array:
+			tv := targetType.(*Array)
+			localType = v.Type
+			targetType = tv.Type
+			continue
+		case *FuncProto:
+			tv := targetType.(*FuncProto)
+			if len(v.Params) != len(tv.Params) {
+				return false, nil
+			}
+
+			for i, p := range v.Params {
+				tp := tv.Params[i]
+				lpType, err := skipModsAndTypedefs(p.Type)
+				if err != nil {
+					return false, err
+				}
+				tpType, err := skipModsAndTypedefs(tp.Type)
+				if err != nil {
+					return false, err
+				}
+				if compat, err := areTypesCompatible(lpType, tpType); !compat || err != nil {
+					return false, err
+				}
+			}
+
+			// tail recurse for return type check
+			localType, err = skipModsAndTypedefs(v.Return)
+			if err != nil {
+				return false, err
+			}
+			targetType, err = skipModsAndTypedefs(tv.Return)
+			if err != nil {
+				return false, err
+			}
+			continue
+		default:
+			return false, nil
+		}
+	}
+
+	return false, errors.New("exceeded type depth")
+}
+
+func areMembersCompatible(localType Type, targetType Type) (bool, error) {
+	var err error
+	for depth := 0; depth <= maxTypeDepth; depth++ {
+		localType, err = skipModsAndTypedefs(localType)
+		if err != nil {
+			return false, err
+		}
+		targetType, err = skipModsAndTypedefs(targetType)
+		if err != nil {
+			return false, err
+		}
+
+		_, lok := localType.(composite)
+		_, tok := targetType.(composite)
+		if lok && tok {
+			return true, nil
+		}
+		if reflect.TypeOf(localType) != reflect.TypeOf(targetType) {
+			return false, nil
+		}
+
+		switch v := localType.(type) {
+		case *Pointer:
+			return true, nil
+		case *Enum:
+			tv := targetType.(*Enum)
+			localEssenName := v.essentialName()
+			targetEssenName := tv.essentialName()
+			matches := localEssenName == "" || targetEssenName == "" || localEssenName == targetEssenName
```

Comment:
Do I understand this correctly: enum members are compatible if either of them is anonymous or if they have the same name?

If that's true, breaking it into two statements seems easier to understand to me:
```
if localEssenName == "" || targetEssenName == "" {
    // Allow changing from anonymous to named member
  return true, nil
}

return localEssenName == targetEssenName nil
```

Reply:
That is my understanding as well.